src/utilities.rs

```diff
@@ -0,0 +1,87 @@
+use std::{collections::BTreeMap, marker::PhantomData};
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(transparent))]
+#[repr(transparent)]
+pub struct Id<T: Clone>(usize, PhantomData<T>);
+
+impl<T> PartialEq for Id<T>
+where
+    T: Clone,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Storage<T: Clone> {
+    next_id: usize,
+    storage: BTreeMap<usize, T>,
+}
+
+impl<T> Storage<T>
+where
+    T: Clone,
+{
+    pub fn new() -> Self {
+        Storage {
+            next_id: 0,
+            storage: BTreeMap::new(),
+        }
+    }
+
+    pub fn add(&mut self, value: T) -> Id<T> {
+        self.storage.insert(self.next_id, value);
+        self.next_id += 1;
+        Id(self.storage.len() - 1, PhantomData::default())
+    }
+
+    // this needs invalidating every Id instance
+    pub fn remove(&mut self, id: Id<T>) -> T {
+        self.storage.remove(&id.0).unwrap()
+    }
+
+    pub fn is_valid(&self, id: &Id<T>) -> bool {
+        self.storage.get(&id.0).is_some()
+    }
+
+    pub fn get(&self, id: &Id<T>) -> &T {
+        // assume the id is valid
+        self.storage.get(&id.0).unwrap()
+    }
+
+    pub fn get_mut(&mut self, id: &Id<T>) -> &mut T {
+        // assume the id is valid
+        self.storage.get_mut(&id.0).unwrap()
+    }
```

Comment:
Wouldn't it be better here to return an Option<T> in order to let the user deal with the case when the id is invalid?

Reply:
You're right, I wanted something more easy to use (or *cheesy*) without having to pile up `if let`s. The remove function arrived late in the process until I figured users may want to delete entries in these storages. If you think it's fine to go with a safer interface I'll change this.

Reply:
Fixed this anyway. I'd prefer having a safe interface at last.